File: problem45/problem45.js
/* Problem 45
 * binarySearch: returns the index of a key in an array. Returns -1 if the key is not in the array
*/
var hex = [];
var n = 50000001;
for (var i = 0; i < n; i++) {
	hex.push(i * (2 * i - 1));
}
var l = 0; var k = 0; var i = 166;
while (true) {
	k = i * (3 * i - 1) / 2;
	if (binarySearch(hex, k, 0, hex.length - 1) > -1) break;
	i++;
}
console.log("The solution is " + k); // every hexagonal number is a triangle number

function binarySearch(T, k, min, max) {
	while (max >= min) {
		var midpoint = Math.floor((max + min) / 2);
		if (T[midpoint] === k) return midpoint;
		else if (T[midpoint] < k) min = midpoint + 1;
		else max = midpoint - 1;
	}
	return -1;
}

File: problem45/Problem45.java
import java.util.*;
/* Problem 45
 * isIn: returns whether a given integer is in a given array where the array is sorted
*/

public class Problem45 {
	public static void main(String[] arg) {
		long[] hex = new long[50000001];
		for (int i = 0; i < hex.length; i++) {
			hex[i] = (long)i * (2 * (long)i - 1);
		}
		long l = 0; long k = 0; int i = 166;
		while (true) {
			l = (long)i;
			k = l * (3 * l - 1) / 2;
			if (isIn(hex, k)) break;
			i++;
		}
		System.out.println("The solution is " + k); // every hexagonal number is a triangle number
	}
	
	public static boolean isIn(long[] T, long t) {	
		int a =  Arrays.binarySearch(T, t);
		return a > 0;
	}
}